### PR TITLE
[Kernel][Predicate] - WIP - Add support for STARTS_WITH expression

### PR DESCRIPTION
**Pull request by krishnanravi**
#### Which Delta project/connector is this regarding?
- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description
addresses  https://github.com/delta-io/delta/issues/2539

## How was this patch tested?
added unit tests

## Does this PR introduce _any_ user-facing changes?
No

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/expressions/StartsWith.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import io.delta.kernel.annotation.Evolving;
+
+/**
+ * {@code STARTS WITH} expression
+ * <p>
+ * Definition:
+ * <p><ul>
+ *     <li>{@code expr1} STARTS WITH {@code expr2} on two inputs.</li>
+ *     <li>Requires both left and right input expressions of type {@link Predicate}.</li>
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/expressions/StartsWith.java:26]: left and right need to be strings, right?

File: kernel/kernel-api/src/main/java/io/delta/kernel/expressions/StartsWith.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ *     <li>Result is null when either of the inputs is null,
+ *     true if left {@code STARTS WITH} right, false otherwise.
+ *     Note that matches are case insensitive.</li>
+ * </ul>
+ *
+ * @since 3.0.0
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/expressions/StartsWith.java:32]: change the version to 3.3.0

**Review by vkorukanti — CHANGES_REQUESTED** [review]
Thank you @krishnanravi for the PR.

Wondering if we can directly support `LIKE`. `STARTS_WITH` is a special form of `LIKE`.

E.g. `STARTS_WITH(column, 'city')` is equals to `column LIKE 'city%'`

File: kernel/kernel-api/src/main/java/io/delta/kernel/expressions/StartsWith.java (hunk 1 above)
Comment (krishnanravi) [reply kernel/kernel-api/src/main/java/io/delta/kernel/expressions/StartsWith.java]: yes. updated.

**Comment by krishnanravi** [comment]
> Thank you @krishnanravi for the PR.
> 
> Wondering if we can directly support `LIKE`. `STARTS_WITH` is a special form of `LIKE`.
> 
> E.g. `STARTS_WITH(column, 'city')` is equals to `column LIKE 'city%'`

@vkorukanti yes makes sense to me. are you suggesting we replace STARTS_WITH with LIKE support instead as part of this change/PR?

**Comment by vkorukanti** [comment]
> @vkorukanti yes makes sense to me. are you suggesting we replace STARTS_WITH with LIKE support instead, and as part of this change/PR?

Yep, lets add the `LIKE` as part of this PR. Basically, replace the `STARTS_WITH` like and the eval to use the regex. There was a similar [PR](https://github.com/delta-io/delta/pull/1953/files) for standalone that you can refer to.  The rest of the changes in this PR are still the same.

**Comment by krishnanravi** [comment]
> > @vkorukanti yes makes sense to me. are you suggesting we replace STARTS_WITH with LIKE support instead, and as part of this change/PR?
> 
> Yep, lets add the `LIKE` as part of this PR. Basically, replace the `STARTS_WITH` like and the eval to use the regex. There was a similar [PR](https://github.com/delta-io/delta/pull/1953/files) for standalone that you can refer to. The rest of the changes in this PR are still the same.

@vkorukanti assume we will need support for a custom escape char in input. pls confirm.

**Comment by vkorukanti** [comment]
> @vkorukanti assume we will need support for a custom escape char in input. e.g. `col1 LIKE '%\%%' ESCAPE '\'` should escape the second % token and the escape char per se can be provided for LiKE predicates.
Let's start with the simple like with no escape and you can follow up the escape in a separate PR. Thanks again for picking this up.

**Pull request closed without merging**